File: ios/Sources/Protocol/Work.swift
// Protocol — the work extension (PROTOCOL.md §10). Parse, serialise, and the vouch format.
// Pure Swift; no platform imports beyond Foundation.
//
// §10 is an extension, and the shape of this file is the argument for it. `Card.swift` is §2 and
// knows nothing about anything here: delete this file and every card on the network still parses
// byte-identically, work keys included, because §2 already says unknown keys are ignored. §10 is
// read by a SECOND pass over the same text.
//
// The two meet in exactly one place — the `CardCodec.serialise(_:work:)` overload at the bottom,
// which emits work lines when, and only when, the caller hands it a `Work`. That is §10.6: a
// client that read the keys has to write them back, or a follow would quietly delete somebody's
// work card.

import Foundation

/// The closed intent set (§10.3). Unknown intents are dropped rather than shown — a client cannot
/// render a word it has no copy for.
public enum WorkIntent: String, Codable, Equatable, Hashable, CaseIterable {
    case work, contract, hiring, collab

    /// The four strings, verbatim (PRODUCT §2.23). Not derived from the raw value: these are copy.
    public var label: String {
        switch self {
        case .work: return "Open to work"
        case .contract: return "Open to contract"
        case .hiring: return "Hiring"
        case .collab: return "Open to collaborate"
        }
    }

    /// The Edit Card tab label (PRODUCT §2.23).
    public var editLabel: String {
        switch self {
        case .work: return "Work"
        case .contract: return "Contract"
        case .hiring: return "Hiring"
        case .collab: return "Collab"
        }
    }
}

/// `<intent> until <YYYY-MM-DD>`, the one time-sensitive thing on a card (§10.3).
public struct WorkOpen: Codable, Equatable, Hashable {
    public var intent: WorkIntent
    /// `YYYY-MM-DD`, interpreted UTC.
    public var until: String

    public init(intent: WorkIntent, until: String) { self.intent = intent; self.until = until }
}

/// The four §10.2 keys, parsed. `nil` from `WorkCodec.parse` means the node has no work card, so
/// "has a work card" is one optional check and a card whose every work line is malformed is the
/// same as a card with none.
public struct Work: Codable, Equatable, Hashable {
    public var role: String?
    public var does: [String]
    public var open: WorkOpen?
    /// Which of the card's own `feeds:` are work (PRODUCT §2.24: the marking is per feed).
    public var feeds: [String]

    public init(role: String? = nil, does: [String] = [], open: WorkOpen? = nil, feeds: [String] = []) {
        self.role = role; self.does = does; self.open = open; self.feeds = feeds
    }

    public var isEmpty: Bool { role == nil && does.isEmpty && open == nil && feeds.isEmpty }

    public func lists(feed username: String) -> Bool {
        let k = Username.key(username)
        return feeds.contains { Username.key($0) == k }
    }
}

public enum WorkCodec {
    public static let roleMax = 80
    public static let doesMax = 12
    public static let tagMin = 2
    public static let tagMax = 24
    /// §10.3: a date further out than this is not a statement about now. `until 2099-01-01` is an
    /// expiry nobody ever has to renew, which is the same as no expiry at all.
    public static let openHorizonDays = 180
    /// PRODUCT §2.23: the horizons the writer offers, well inside the cap.
    public static let openHorizons = [30, 60, 90]

    static let keys: Set<String> = ["work.role", "work.does", "work.open", "work.feeds"]

    // MARK: Tags (§10.2)

    /// One capability tag, normalised: trimmed, inner whitespace collapsed, lowercased. Invalid → nil.
    ///
    /// The grammar is `[a-z0-9][a-z0-9 +#.-]{0,22}[a-z0-9+#]` — the punctuation real trades carry,
    /// so `c++`, `c#`, `node.js` and `front of house` are tags and `live/sound` is not. Written out
    /// rather than as a regex because every character class here is a decision worth reading.
    public static func tag(_ input: String) -> String? {
        var collapsed = ""
        var lastWasSpace = false
        for ch in input.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
            let isSpace = ch == " " || ch == "\t" || ch == "\n"
            if isSpace {
                if !lastWasSpace { collapsed.append(" ") }
                lastWasSpace = true
            } else {
                collapsed.append(ch)
                lastWasSpace = false
            }
        }
        let chars = Array(collapsed)
        guard chars.count >= tagMin, chars.count <= tagMax else { return nil }
        guard isEdgeStart(chars[0]), isEdgeEnd(chars[chars.count - 1]) else { return nil }
        for ch in chars.dropFirst().dropLast() where !isInner(ch) { return nil }
        return collapsed
    }

    private static func isEdgeStart(_ ch: Character) -> Bool { ch.isASCII && (ch.isLowercase || ch.isNumber) }
    private static func isEdgeEnd(_ ch: Character) -> Bool { isEdgeStart(ch) || ch == "+" || ch == "#" }
    private static func isInner(_ ch: Character) -> Bool { isEdgeEnd(ch) || ch == " " || ch == "." || ch == "-" }

    /// `a, b, c` → up to `doesMax` tags. Invalid tags are dropped, never fatal (§10.2); duplicates
    /// collapse to the first.
    public static func does(from value: String) -> [String] {
        var out: [String] = []
        var seen = Set<String>()
        for part in value.split(separator: ",", omittingEmptySubsequences: false) {
            guard let t = tag(String(part)), seen.insert(t).inserted else { continue }
            out.append(t)
            if out.count == doesMax { break }
        }
        return out
    }

    // MARK: Dates (§10.3)

    /// A real calendar day in `YYYY-MM-DD` — `2026-02-30` is not one.
    public static func isCalendarDay(_ s: String) -> Bool { civil(s) != nil }

    /// Whole days from `a` to `b`, both `YYYY-MM-DD`. Negative when `b` is behind `a`; nil when
    /// either is not a day. Integer arithmetic on the proleptic Gregorian calendar rather than
    /// `Calendar`, because §10.3 is UTC and a reader's time zone must not move an expiry.
    public static func daysBetween(_ a: String, _ b: String) -> Int? {
        guard let x = civil(a), let y = civil(b) else { return nil }
        return daysFromCivil(y) - daysFromCivil(x)
    }

    /// Today as `YYYY-MM-DD`, UTC — the `today` every reader-side expiry check takes.
    public static func today(_ now: Date = Date()) -> String {
        day(from: Int(floor(now.timeIntervalSince1970 / 86_400)))
    }

    /// `<intent> until <YYYY-MM-DD>` → the open intent; anything else → nil (§10.3).
    public static func open(from value: String) -> WorkOpen? {
        let parts = value.trimmingCharacters(in: .whitespaces)
            .split(whereSeparator: { $0 == " " || $0 == "\t" })
        guard parts.count == 3, parts[1].lowercased() == "until" else { return nil }
        guard parts[0].allSatisfy({ $0.isASCII && $0.isLetter }) else { return nil }
        guard let intent = WorkIntent(rawValue: parts[0].lowercased()) else { return nil }
        let until = String(parts[2])
        guard isCalendarDay(until) else { return nil }
        return WorkOpen(intent: intent, until: until)
    }

    public static func serialise(open: WorkOpen?) -> String {
        guard let open else { return "" }
        return "\(open.intent.rawValue) until \(open.until)"
    }

    /// Is this intent a statement about now? Expired is obvious; the far-future cap is the other
    /// half, and it is the one that makes the first rule mean something (§10.3). `today` is
    /// `YYYY-MM-DD`.
    public static func isCurrent(_ open: WorkOpen?, today: String) -> Bool {
        guard let open, let days = daysBetween(today, open.until) else { return false }
        return days >= 0 && days <= openHorizonDays
    }

    /// `YYYY-MM-DD` `horizon` days after `today` — what the writer's `FOR` tabs mint (PRODUCT §2.23).
    public static func day(after horizon: Int, from today: String = WorkCodec.today()) -> String? {
        guard let c = civil(today) else { return nil }
        return day(from: daysFromCivil(c) + horizon)
    }

    /// The `FOR` tab that covers `left` days (PRODUCT §2.23). A card carries the end DATE, never
    /// the horizon that produced it, so reopening Edit Card has to recover one — and a control
    /// showing a horizon the card does not have is not cosmetic: the next Save writes it, so an
    /// edit to the bio alone would move the one date §10.3 makes the reader enforce.
    public static func horizon(remainingDays left: Int) -> Int {
        openHorizons.first { $0 >= left } ?? openHorizons.last ?? left
    }

    // MARK: The card pass (§10.2)

    /// The §10 pass over a card's text. Returns the work card, or nil when the text is not a v1
    /// card or carries nothing §10 can use.
    ///
    /// `work.feeds` is intersected with `feeds:` here rather than trusted: `feeds:` is the
    /// ownership claim (§3, which needs post rights), and a marking line has no business
    /// introducing a channel the owner never claimed.
    public static func parse(_ text: String) -> Work? {
        guard let card = CardCodec.parse(text).card else { return nil }
        let lines = text.replacingOccurrences(of: "\r\n", with: "\n").replacingOccurrences(of: "\r", with: "\n")
            .split(separator: "\n", omittingEmptySubsequences: false).map(String.init)
        var raw: [String: String] = [:]
        for line in lines.dropFirst() {
            guard let colon = line.firstIndex(of: ":") else { continue }
            let key = line[..<colon].trimmingCharacters(in: .whitespaces).lowercased()
            guard keys.contains(key) else { continue }
            let value = line[line.index(after: colon)...].trimmingCharacters(in: .whitespaces)
            // §2's repetition rule, unchanged: a repeated key concatenates with a space.
            if let existing = raw[key], !existing.isEmpty {
                raw[key] = value.isEmpty ? existing : existing + " " + value
            } else {
                raw[key] = value
            }
        }
        var work = Work()
        work.role = role(from: raw["work.role"] ?? "")
        work.does = does(from: raw["work.does"] ?? "")
        work.open = open(from: raw["work.open"] ?? "")
        work.feeds = Username.list(from: raw["work.feeds"] ?? "").filter { card.lists(feed: $0) }
        return work.isEmpty ? nil : work
    }

    /// One line, whitespace collapsed, capped. A longer role keeps its first 80 characters and
    /// never rejects the card (§10.2).
    static func role(from value: String) -> String? {
        let collapsed = value.split(whereSeparator: { $0 == " " || $0 == "\t" || $0 == "\n" }).joined(separator: " ")
        return collapsed.isEmpty ? nil : String(collapsed.prefix(roleMax))
    }

    /// The §10 lines, in §10.2 order, for `CardCodec.serialise(_:work:)` to append after `replies`.
    /// Every value is re-normalised on the way out, so a card written from a hand-built `Work` is
    /// the same bytes as one round-tripped from the network.
    ///
    /// `feeds` is the card's own `feeds:`. §10.2 says `work.feeds` is a marking on channels the card
    /// already claims, so the intersection is enforced on the write side here exactly as `parse`
    /// enforces it on the read side: no caller can put an entry on the wire the card does not list.
    /// It matters because a stale marking is latent, not harmless — readers ignore it today and it
    /// re-marks the channel as work the moment the owner lists it again, with nobody touching the
    /// toggle.
    public static func lines(_ work: Work?, feeds: [String] = []) -> [String] {
        guard let work else { return [] }
        var out: [String] = []
        if let role = role(from: work.role ?? "") { out.append("work.role: " + role) }
        let tags = does(from: work.does.joined(separator: ","))
        if !tags.isEmpty { out.append("work.does: " + tags.joined(separator: ", ")) }
        if let open = open(from: serialise(open: work.open)) { out.append("work.open: " + serialise(open: open)) }
        let marked = Username.list(from: work.feeds.map { $0.hasPrefix("@") ? $0 : "@" + $0 }.joined(separator: " "))
            .filter { m in feeds.contains { Username.key($0) == Username.key(m) } }
        if !marked.isEmpty { out.append("work.feeds: " + marked.map { "@" + $0 }.joined(separator: " ")) }
        return out
    }

    // MARK: Civil date arithmetic

    /// `YYYY-MM-DD` → (y, m, d), nil when it is not a real calendar day.
    static func civil(_ s: String) -> (y: Int, m: Int, d: Int)? {
        let chars = Array(s)
        guard chars.count == 10, chars[4] == "-", chars[7] == "-" else { return nil }
        for (i, ch) in chars.enumerated() where i != 4 && i != 7 {
            guard ch.isASCII, ch.isNumber else { return nil }
        }
        guard let y = Int(String(chars[0..<4])), let m = Int(String(chars[5..<7])), let d = Int(String(chars[8..<10])) else { return nil }
        guard m >= 1, m <= 12, d >= 1, d <= daysIn(month: m, year: y) else { return nil }
        return (y, m, d)
    }

    static func daysIn(month: Int, year: Int) -> Int {
        switch month {
        case 1, 3, 5, 7, 8, 10, 12: return 31
        case 4, 6, 9, 11: return 30
        default: return isLeap(year) ? 29 : 28
        }
    }

    static func isLeap(_ y: Int) -> Bool { (y % 4 == 0 && y % 100 != 0) || y % 400 == 0 }

    /// Days from 1970-01-01. Howard Hinnant's `days_from_civil`, which is exact for every year the
    /// proleptic Gregorian calendar covers and needs no calendar object.
    static func daysFromCivil(_ c: (y: Int, m: Int, d: Int)) -> Int {
        let y = c.y - (c.m <= 2 ? 1 : 0)
        let era = (y >= 0 ? y : y - 399) / 400
        let yoe = y - era * 400
        let doy = (153 * (c.m + (c.m > 2 ? -3 : 9)) + 2) / 5 + c.d - 1
        let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy
        return era * 146_097 + doe - 719_468
    }

    /// The inverse: days from 1970-01-01 → `YYYY-MM-DD`.
    static func day(from days: Int) -> String {
        let z = days + 719_468
        let era = (z >= 0 ? z : z - 146_096) / 146_097
        let doe = z - era * 146_097
        let yoe = (doe - doe / 1_460 + doe / 36_524 - doe / 146_096) / 365
        let y = yoe + era * 400
        let doy = doe - (365 * yoe + yoe / 4 - yoe / 100)
        let mp = (5 * doy + 2) / 153
        let d = doy - (153 * mp + 2) / 5 + 1
        let m = mp + (mp < 10 ? 3 : -9)
        let year = y + (m <= 2 ? 1 : 0)
        return String(format: "%04d-%02d-%02d", year, m, d)
    }
}

// MARK: - The vouch (§10.4)

/// One vouch as written: the subject node, the one capability, and an optional body.
public struct VouchBody: Equatable, Hashable {
    public var node: String
    public var does: String
    public var body: String
}

public enum VouchCodec {
    public static let prefix = "vouch: "
    public static let doesPrefix = "does: "
    static let linkHost = "https://t.me/"

    /// §10.4, exactly. Both lines are mandatory: a `vouch:` with no `does:` would assert "I vouch
    /// for this person", which nobody can weigh and which decays into a like button inside a week.
    /// A message missing either line is not a vouch and readers skip it, the same way §6.2 skips a
    /// message with no `re:` line.
    public static func parse(_ text: String) -> VouchBody? {
        let lines = text.split(separator: "\n", omittingEmptySubsequences: false).map {
            $0.hasSuffix("\r") ? String($0.dropLast()) : String($0)
        }
        guard lines.count >= 2 else { return nil }
        guard let node = target(lines[0]) else { return nil }
        guard lines[1].hasPrefix(doesPrefix) else { return nil }
        let rest = String(lines[1].dropFirst(doesPrefix.count))
        guard !rest.isEmpty, let does = WorkCodec.tag(rest) else { return nil }
        let body = lines.count > 2 ? lines[2...].joined(separator: "\n") : ""
        return VouchBody(node: node, does: does, body: body)
    }

    /// Line 1: `vouch: ` — one space — then the NODE channel's link, no message id and no trailing
    /// slash. A link with a message id is a §6.2 comment, not a vouch, and this is the line that
    /// keeps the two apart in one channel.
    static func target(_ line: String) -> String? {
        guard line.hasPrefix(prefix) else { return nil }
        var rest = String(line.dropFirst(prefix.count))
        while rest.hasSuffix(" ") || rest.hasSuffix("\t") { rest.removeLast() }
        guard rest.hasPrefix(linkHost) else { return nil }
        let name = String(rest.dropFirst(linkHost.count))
        guard Username.isValid(name) else { return nil }
        return name
    }

    /// §10.4, exact bytes. Nil when the tag is one §10.2 would drop — the caller's refusal is
    /// `Pick one thing.` (PRODUCT §2.25).
    public static func serialise(node: String, does: String, body: String) -> String? {
        guard let tag = WorkCodec.tag(does), let name = Username.normalise(node) else { return nil }
        let head = prefix + linkHost + name + "\n" + doesPrefix + tag
        return body.isEmpty ? head : head + "\n" + body
    }

    /// §10.4: a vouch for the channel's own owner is not a vouch. Unforgeable-by-construction is
    /// the whole value of the format, and it holds only because the one channel a person can write
    /// is the one that cannot speak about them. A client that renders a self-vouch has given that
    /// away, so the check sits at the parse boundary and a self-vouch never reaches an index.
    public static func keeps(_ vouch: VouchBody?, voucherNode: String) -> Bool {
        guard let vouch, !voucherNode.isEmpty else { return false }
        return Username.key(vouch.node) != Username.key(voucherNode)
    }
}

// MARK: - Where §10 meets §2 (§10.6)

public extension CardCodec {
    /// §2's serialiser with the §10 lines appended, in §10.2 order, each omitted when empty. §2's
    /// own order and output are unchanged, so a card written before this section and one written
    /// after differ by an append.
    ///
    /// This is the ONE place the extension touches the protocol's own codec, and it is the whole of
    /// §10.6: a client that implements §10 must write back the work lines it read, or a follow —
    /// which rewrites the entire pinned message — quietly deletes somebody's work card.
    static func serialise(_ card: Card, work: Work?) -> String {
        let lines = WorkCodec.lines(work, feeds: card.feeds)
        let base = serialise(card)
        return lines.isEmpty ? base : base + "\n" + lines.joined(separator: "\n")
    }

    /// §2's 4096 cap, unmoved, now counting the work lines the same write would carry.
    static func isFull(_ card: Card, work: Work?) -> Bool { serialise(card, work: work).count > maxLength }
}

// MARK: - Work dates on screen (PRODUCT §2.23, §2.25)

public enum WorkDate {
    /// `until 1 Dec` — day and month, the year appended only when it is not this one (§2.23).
    /// Derived from the date, never typed.
    public static func short(_ day: String, now: Date = Date(), calendar: Calendar = .current) -> String? {
        guard let c = WorkCodec.civil(day) else { return nil }
        let thisYear = calendar.component(.year, from: now)
        return "\(c.d) \(monthAbbrev(c.m))" + (c.y == thisYear ? "" : " \(c.y)")
    }

    /// `5 Dec 2026` — the Edit Card `FOR` line, which always carries the year because it is a date
    /// the writer is choosing rather than reading.
    public static func full(_ day: String) -> String? {
        guard let c = WorkCodec.civil(day) else { return nil }
        return "\(c.d) \(monthAbbrev(c.m)) \(c.y)"
    }

    /// `Mar 2026` — a vouch's date (§2.25). Everywhere else in this app time is relative, because a
    /// post's recency is what matters; a vouch is the opposite, and `2y ago` buries exactly the
    /// thing a reader is weighing.
    public static func monthYear(unix: Int, calendar: Calendar = .current) -> String {
        let date = Date(timeIntervalSince1970: TimeInterval(unix))
        let m = calendar.component(.month, from: date)
        let y = calendar.component(.year, from: date)
        return "\(monthAbbrev(m)) \(y)"
    }

    static func monthAbbrev(_ m: Int) -> String {
        let names = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]
        guard m >= 1, m <= 12 else { return "" }
        return names[m - 1]
    }
}

File: android/app/src/main/kotlin/ca/lucianlabs/tgsocial/protocol/Work.kt
package ca.lucianlabs.tgsocial.protocol

import kotlinx.serialization.Serializable
import java.time.LocalDate
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.time.format.ResolverStyle
import java.time.temporal.ChronoUnit
import java.util.Locale

/**
 * PROTOCOL §10.3 — expiring intent. [intent] is one of [WorkFormat.INTENTS]; [until] is `YYYY-MM-DD`, UTC.
 * Nothing here records when it was written: both reader rules are arithmetic on this date alone.
 */
@Serializable
data class WorkOpen(val intent: String, val until: String)

/**
 * PROTOCOL §10.2 — the work card: four optional keys, all self-claims, none of them checkable by anyone
 * (§10.8). It hangs off [Card] rather than replacing anything on it, which is what makes §10 an extension
 * instead of a version.
 */
@Serializable
data class WorkCard(
    val role: String? = null,
    val does: List<String> = emptyList(),
    val open: WorkOpen? = null,
    /** Feeds of mine that are work. Every entry also appears in the card's own `feeds:` (§10.2). */
    val feeds: List<String> = emptyList(),
) {
    val isEmpty: Boolean get() = role == null && does.isEmpty() && open == null && feeds.isEmpty()

    fun marks(feed: String): Boolean = feeds.any { Username.same(it, feed) }
}

/**
 * PROTOCOL §10 — the work extension, read as a SECOND pass over the same bytes [CardFormat] already read.
 *
 * [CardFormat] knows nothing about any of this and is not touched by it: `work.role` carries a `.`, which
 * §2's key filter already rejects, so a card with these lines parses into exactly the card it parsed before
 * this file existed. Delete this file and the network still works — that is §10's whole claim, and
 * `docs/card-vectors.json` runs it as a test through the §2 loop every client already has.
 *
 * The two meet in one place: [CardFormat.serialise] emits these lines when, and only when, it is handed a
 * [WorkCard]. That is §10.6 — a client that read the keys has to write them back, or the next follow
 * silently deletes somebody's work card.
 */
object WorkFormat {
    const val ROLE_MAX = 80
    const val DOES_MAX = 12

    /**
     * §10.3 — a date further out than this is not a statement about now. Without it `until 2099-01-01` is an
     * expiry nobody ever has to renew, which is the same as no expiry, which is what the key exists to avoid.
     */
    const val OPEN_HORIZON_DAYS = 180L

    /** §10.3 — closed set. A client cannot render a word it has no copy for, so an unknown intent is absent. */
    val INTENTS: List<String> = listOf("work", "contract", "hiring", "collab")

    /** §10.2 — lowercase, 2–24, and the punctuation real trades carry: `c++`, `c#`, `node.js`, `front of house`. */
    private val TAG = Regex("^[a-z0-9][a-z0-9 +#.-]{0,22}[a-z0-9+#]$")
    private val WORK_KEYS = setOf("work.role", "work.does", "work.open", "work.feeds")
    private val OPEN = Regex("^([A-Za-z]+)\\s+until\\s+(\\S+)$")
    private val WHITESPACE = Regex("\\s+")
    /**
     * §10.2 — STRICT, so `2026-02-30` is not a day. The lenient resolver quietly folds it back to the 28th,
     * which would make a card carrying an impossible date render an intent that expires on a date nobody
     * wrote — the value is malformed and the key is absent, which is what the vector asserts.
     */
    private val DAY = DateTimeFormatter.ofPattern("uuuu-MM-dd", Locale.ROOT).withResolverStyle(ResolverStyle.STRICT)

    // ---------------------------------------------------------------- tags

    /** One capability tag, normalised: trimmed, inner whitespace collapsed, lowercased. Invalid → null. */
    fun tag(input: String?): String? {
        val s = input?.trim()?.replace(WHITESPACE, " ")?.lowercase() ?: return null
        return if (TAG.matches(s)) s else null
    }

    /** `a, b, c` → up to [DOES_MAX] tags. An invalid tag is dropped and the rest of the line stands (§10.2). */
    fun does(value: String?): List<String> {
        if (value.isNullOrBlank()) return emptyList()
        val out = ArrayList<String>(DOES_MAX)
        val seen = HashSet<String>()
        for (part in value.split(',')) {
            val tag = tag(part) ?: continue
            if (!seen.add(tag)) continue
            out += tag
            if (out.size == DOES_MAX) break
        }
        return out
    }

    // ---------------------------------------------------------------- dates

    /** A real calendar day in `YYYY-MM-DD`. `2026-02-30` matches the shape and is not one. */
    fun isCalendarDay(s: String?): Boolean = day(s) != null

    private fun day(s: String?): LocalDate? {
        if (s == null || s.length != 10) return null
        return try {
            LocalDate.parse(s, DAY)
        } catch (_: DateTimeParseException) {
            null
        }
    }

    /** Whole days from [a] to [b], both `YYYY-MM-DD`. Negative when [b] is behind [a]. */
    fun daysBetween(a: String, b: String): Long {
        val from = day(a) ?: return 0
        val to = day(b) ?: return 0
        return ChronoUnit.DAYS.between(from, to)
    }

    /** §10.3 — today, UTC, in the form the card writes. */
    fun today(): String = LocalDate.now(ZoneOffset.UTC).format(DAY)

    /** `<intent> until <YYYY-MM-DD>` → [WorkOpen]; anything else, including an unknown intent, → null. */
    fun parseOpen(value: String?): WorkOpen? {
        val m = OPEN.find(value?.trim().orEmpty()) ?: return null
        val intent = m.groupValues[1].lowercase()
        if (intent !in INTENTS) return null
        val until = m.groupValues[2]
        if (!isCalendarDay(until)) return null
        return WorkOpen(intent, until)
    }

    fun serialiseOpen(open: WorkOpen?): String = open?.let { "${it.intent} until ${it.until}" }.orEmpty()

    /**
     * §10.3 — is this intent a statement about *now*? Expired is the obvious half; the horizon is the other,
     * and it is the one that makes the first mean anything. Pure arithmetic on the card text: no write date
     * exists to compare against, and none is needed.
     */
    fun openIsCurrent(open: WorkOpen?, today: String = today()): Boolean {
        if (open == null || !isCalendarDay(open.until) || !isCalendarDay(today)) return false
        val days = daysBetween(today, open.until)
        return days >= 0 && days <= OPEN_HORIZON_DAYS
    }

    /** Days from today until [open] ends — negative once it has. Null when there is no readable date. */
    fun daysLeft(open: WorkOpen?, today: String = today()): Long? {
        if (open == null || !isCalendarDay(open.until)) return null
        return daysBetween(today, open.until)
    }

    // ---------------------------------------------------------------- the card

    /**
     * The §10 pass over a card's text. Null when the text is not a v1 card, or carries nothing §10 can use —
     * so "has a work card" is one null check, and a card whose every work line is malformed is the same as a
     * card with none (§10.2).
     *
     * `work.feeds` is intersected with `feeds:` here rather than trusted: `feeds:` is the ownership claim
     * (§3 requires post rights for it), so a marking line has no business introducing a channel the owner
     * never claimed.
     */
    fun parse(text: String): WorkCard? {
        val card = (CardFormat.parse(text) as? CardParse.Parsed)?.card ?: return null
        val raw = LinkedHashMap<String, StringBuilder>()
        val lines = text.split("\n").map { it.trimEnd('\r') }
        for (line in lines.drop(1)) {
            val colon = line.indexOf(':')
            if (colon <= 0) continue
            val key = line.substring(0, colon).trim().lowercase()
            if (key !in WORK_KEYS) continue
            val value = line.substring(colon + 1).trim()
            // §10.2 — repetition follows §2: a repeated key concatenates with a space.
            val sb = raw.getOrPut(key) { StringBuilder() }
            if (sb.isNotEmpty() && value.isNotEmpty()) sb.append(' ')
            sb.append(value)
        }
        val role = raw["work.role"]?.toString()?.replace(WHITESPACE, " ")?.trim().orEmpty()
        val work = WorkCard(
            role = role.takeIf { it.isNotEmpty() }?.take(ROLE_MAX),
            does = does(raw["work.does"]?.toString()),
            open = parseOpen(raw["work.open"]?.toString()),
            feeds = Username.list(raw["work.feeds"]?.toString().orEmpty()).filter { card.hasFeed(it) },
        )
        return work.takeIf { !it.isEmpty }
    }

    /**
     * The §10 pass attached to a card the caller already parsed with §2's own parser — the one call site that
     * joins the two passes on the read side, so §10.6's "write back what you read" holds for every card that
     * enters the app rather than for the ones somebody remembered.
     */
    fun attach(card: Card?, text: String?): Card? {
        if (card == null) return null
        return card.copy(work = text?.let { parse(it) })
    }

    /** The §10 lines, in §10.2 order, for [CardFormat.serialise] to append after `replies`. */
    /**
     * [feeds] is the card's own `feeds:`. §10.2 says `work.feeds` is a marking on channels the card
     * already claims, so the intersection is enforced on the write side here exactly as [parse] enforces
     * it on the read side: no caller can put an entry on the wire the card does not list. A stale marking
     * is latent rather than harmless — readers ignore it today and it re-marks the channel as work the
     * moment the owner lists it again, with nobody touching the toggle.
     */
    fun lines(work: WorkCard?, feeds: List<String> = emptyList()): List<String> {
        if (work == null) return emptyList()
        val out = ArrayList<String>(4)
        val role = work.role?.replace(WHITESPACE, " ")?.trim()?.take(ROLE_MAX).orEmpty()
        if (role.isNotEmpty()) out += "work.role: $role"
        val does = does(work.does.joinToString(","))
        if (does.isNotEmpty()) out += "work.does: ${does.joinToString(", ")}"
        val open = parseOpen(serialiseOpen(work.open))
        if (open != null) out += "work.open: ${serialiseOpen(open)}"
        val marked = Username.list(work.feeds.joinToString(" ") { "@${it.removePrefix("@")}" })
            .filter { m -> feeds.any { Username.same(it, m) } }
        if (marked.isNotEmpty()) out += "work.feeds: ${marked.joinToString(" ") { "@$it" }}"
        return out
    }

    // ---------------------------------------------------------------- display (PRODUCT §2.23)

    /** The four intent strings, verbatim (PRODUCT §2.23). An intent outside the set has no copy and no pill. */
    fun intentLabel(intent: String?): String? = when (intent) {
        "work" -> "Open to work"
        "contract" -> "Open to contract"
        "hiring" -> "Hiring"
        "collab" -> "Open to collaborate"
        else -> null
    }

    /**
     * PRODUCT §2.23 — `until 1 Dec`: day and month, derived, the year appended only when it is not this one.
     * Derived rather than recalled, like every other date this app prints.
     */
    fun untilLabel(until: String, today: String = today(), locale: Locale = Locale.getDefault()): String? {
        val end = day(until) ?: return null
        val now = day(today) ?: return null
        val pattern = if (end.year == now.year) "d MMM" else "d MMM yyyy"
        return "until " + DateTimeFormatter.ofPattern(pattern, locale).format(end)
    }

    /**
     * PRODUCT §2.25 — a vouch's date is a month and a year. Everywhere else time is relative because a post's
     * recency is what matters; here `2y ago` buries the thing the reader is weighing.
     */
    fun monthYear(epochSeconds: Long, locale: Locale = Locale.getDefault()): String =
        DateTimeFormatter.ofPattern("MMM yyyy", locale)
            .format(java.time.LocalDateTime.ofInstant(java.time.Instant.ofEpochSecond(epochSeconds), java.time.ZoneId.systemDefault()))

    /** PRODUCT §2.23 — the writer's three horizons, as day counts. */
    val HORIZONS: List<Int> = listOf(30, 60, 90)

    /**
     * Which `FOR` tab the Edit Card modal opens on for a card that already carries intent: the shortest
     * horizon that still covers what is left of it.
     *
     * Seeded, not defaulted, because the tabs are also what Save writes. A modal that opens on `30 days`
     * over a card reading `until <today+90>` reports a horizon the card does not have, contradicts the pill
     * on the profile next to it, and then makes itself true on the next Save — shortening published intent
     * (§10.3) with no refusal and no note, triggered by editing the bio.
     */
    fun horizonFor(open: WorkOpen?, today: String = today()): Int {
        val left = daysLeft(open, today) ?: return HORIZONS.first()
        return HORIZONS.firstOrNull { it >= left } ?: HORIZONS.last()
    }

    /** `today + days`, in card form — what the FOR tabs write and what `Ends 5 Dec 2026.` is derived from. */
    fun horizonDate(days: Int, today: String = today()): String {
        val now = day(today) ?: LocalDate.now(ZoneOffset.UTC)
        return now.plusDays(days.toLong()).format(DAY)
    }

    /** `Ends 5 Dec 2026.` — the faint line under FOR (PRODUCT §2.23); the date is derived, never typed. */
    fun endsLabel(until: String, locale: Locale = Locale.getDefault()): String? {
        val end = day(until) ?: return null
        return "Ends " + DateTimeFormatter.ofPattern("d MMM yyyy", locale).format(end) + ". After that it stops showing."
    }
}
